File: programmers/258707.cpp
#include <string>
#include <vector>

using namespace std;

#define N 1010

int n, cur, cnt;
bool chk[N];
int val[N];

int solution(int coin, vector<int> cards) {
    int answer = 0;
    
    n = cards.size();
    for (int i = n; i >= 0; i--) {
        if (val[i]) continue;
        for (int j = i - 1; j >= 0; j--) {
            if (cards[i] + cards[j] == n + 1) {
                val[i] = j, val[j] = i;
            }
        }
    }
    
    cur = n / 3;
    for (int i = 0; i < n / 3; i++) {
        chk[i] = true;
        if (chk[i] && chk[val[i]]) cnt++;
    }
    
    answer = 1;
    while (true) {
        
        // 코인 1개 소모하는 경우
        while (cur < n) {
            for (int i = cur; i <= cur + 1; i++) {
                if (chk[i]) continue;
                if (coin && chk[val[i]]) {
                    coin--;
                    chk[i] = true;
                    cnt++;
                }
            }
            
            if (cnt) cnt--;
            else break;
            
            cur += 2;
            answer++;
        }
        
        if (coin < 2 || cur >= n) break;
        
        // 코인 2개 소모하는 경우
        for (int i = cur + 1; i >= n / 3; i--) {
            if (chk[i]) continue;
            if (val[i] < i) {
                coin -= 2;
                chk[i] = chk[val[i]] = true;
                cnt++;
                break;
            }
        }
        
        if (cnt == 0) break;
    }
    
    return answer;
}

File: programmers/258707.java
import java.util.*;

class Solution {
    public int solution(int coin, int[] cards) {
        int n = cards.length;
        
        // 카드 값 -> 인덱스 매핑
        int[] valToIdx = new int[n + 1];
        for (int i = 0; i < n; i++) {
            valToIdx[cards[i]] = i;
        }
        
        boolean[] used = new boolean[n];
        int maxIdx = n / 3 + 2;
        int rounds = 1;
        
        // 라운드 진행
        boolean updated = true;
        while (updated) {
            updated = false;
            
            for (int i = 0; i < maxIdx; i++) {
                if (used[i]) continue;
                
                // 짝이 되는 카드 인덱스
                int j = valToIdx[n + 1 - cards[i]];
                if (maxIdx <= j) continue;
                
                // 카드를 뽑는 데 드는 비용
                int cost = (n / 3 <= i ? 1 : 0) + (n / 3 <= j ? 1 : 0);
                if (coin < cost) continue;
                
                coin -= cost;
                used[i] = used[j] = true;
                
                maxIdx = Math.min(maxIdx + 2, n);
                rounds++;
                updated = true;
                break;
            }
        }
        
        // 가능한 최대 라운드 수
        int maxRounds = (n - n / 3) / 2 + 1;
        return Math.min(rounds, maxRounds);
    }
}